projectq/ops/_qaagate.py

```diff
@@ -0,0 +1,31 @@
+#   Copyright 2019 ProjectQ-Framework (www.projectq.ch)
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+from ._basics import BasicGate
+
+
+class QAA(BasicGate):
+    """
+    Quantum Aplitude Aplification gate.
+
+    See setups.decompositions for the complete implementation
+    """
+    def __init__(self, algorithm, algorithm_inverse, oracle):
```

Comment:
This is a good question. I was thinking on this previous to the creation of the PR. Is it possible to generate automatically the inverse in any case (a general algorithm implemented in a function)? I think it could be very challenging. If the `algorithm` is as simple as one gate (H for example it could be easy, but in the case that `algorithm` is made out of several multi-qubit gates with rotations, S and T gates, IMHO it could be very difficult to generate the inverse algorithm.

I would like to hear what do you think about this and if you have some idea on how to do this. You named that the compiler would take care of generating the inverse. How? I don't know how the compiler can take care of this task. (I'm sure I'm missing something)

Any help will be welcome

Reply:
Actually, ProjectQ already knows how to automatically generate the inverse of a list of instructions: simply use the `Compute`/`Uncompute` context managers ([see documentation here](https://projectq.readthedocs.io/en/latest/projectq.meta.html#projectq.meta.Compute))

Knowing that these exist, are there any other reason why this could not be possible ?

Reply:
I know the Compute/Uncompute but may be I need some help on how to use it in this context.

In fact I was thinking on using this feature for AA. But as per the algorithm definition you would need to use an Uncopute previous to any Compute:

From the reference https://arxiv.org/pdf/quant-ph/0005055.pdf , the expresion (1) presents

Q=Q(A, χ) =−AS0A^(−1)Sχ.

With "A" refering to the algorithm. As you can see the inverse of the algorithm is executed before the algorithm itself (inside the AA).

A possibility is to ask for the inverse of the algorithm and Uncompute to execute the Algorithm (A). I was thinking on this as well. My concern with this option is that form an user perspective he/she has to prepare both in advance, therefore it could be easy to ask for both provided that the user have both.

What do you think?

Reply:
Actually, looking at the code for `Compute`/`Uncompute`, there are essentially three cases we need to consider for this:
  1. no qubit allocation/deallocation within the `algorithm` function
  2. qubit allocation ***but*** no qubit deallocation within the `algorithm` function
  3. qubit allocation ***and*** qubit deallocation within the `algorithm` function

Case 1 can be simply handled by using the `Dagger` context manager (as long as we can ensure that no qubit allocation/deallocation occurs within`algorithm`). For the other two, I still need to think of a good way of doing this.

Reply:
Thanks a lot.

I was assuming that only Case 1 was possible, provided that the algorithm acts only on the system qubits (no ancilla qubits are required and therefore no qubit allocation is required into the algorithm), and that (as requested by the supporting literature) "Let A be any quantum algorithm that acts on H (Hilbert space) and uses no measurements" and therefore no qubit deallocation is required.

It is true that if we think in a very generic algorithm, it could include the use of ancillas that could be allocated inside the algorithm, but we could restrict aceptable algorithms to those that do not require ancillas inside, and therefore we could stick to Case 1.

In that case I would need to do some tests in order to verify how Dagger works and implement the AA without requesting the inverse of the algorithm.

What do you think?

Reply:
Using `Dagger` is really simple:

```python
with Dagger(eng):
    algorithm(system_qubits)
```
     
would generate the inverse of `algorithm` automatically; essentially, storing all the commands within the new context in a list and then sending their inverse them to the next engine in reverse order.

Reply:
Thanks,

I will do some tests and will change the code to ask just for the algorithm, creating the inverse with Dagger.

I will resolve the comment once I have this done